File: onegene/onegene/report/ot_report/ot_report.js
// Copyright (c) 2024, TEAMPRO and contributors
// For license information, please see license.txt

frappe.query_reports["OT Report"] = {
	"filters": [
		{

			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 1,
			
		},
		{
				"fieldname":"to_date",
				"label": __("To Date"),
				"fieldtype": "Date",
				"reqd": 1,
				
			},
			{
				"fieldname": "employee",
				"label": __("Employee"),
				"fieldtype": "Link",
				"options": "Employee"
			},	
	],
	onload: function (report) {
		var to_date = frappe.query_report.get_filter('to_date');
		to_date.refresh();
		var from_date = frappe.query_report.get_filter('from_date');
		from_date.refresh();		var previous_month_end_date = frappe.datetime.add_months(frappe.datetime.month_end(),0);
    	to_date.set_input(previous_month_end_date);
		var previous_month_start_date = frappe.datetime.add_months(frappe.datetime.month_start(),0);
    	from_date.set_input(previous_month_start_date);
	}
};

File: onegene/onegene/report/ot_report/ot_report.py
# Copyright (c) 2024, TEAMPRO and contributors
# For license information, please see license.txt

# import frappe


from __future__ import unicode_literals
from six import string_types
import frappe
import json
import datetime
from datetime import datetime
from frappe.utils import (getdate, cint, add_months, date_diff, add_days,
	nowdate, get_datetime_str, cstr, get_datetime, now_datetime, format_datetime,format_date)
from calendar import monthrange
from frappe import _, msgprint
from frappe.utils import flt
from frappe.utils import cstr, cint, getdate
from itertools import count
import datetime as dt
from datetime import datetime, timedelta



def execute(filters=None):
	data = []
	columns = get_columns()
	otreport = get_ot(filters)
	for ot in otreport:
		data.append(ot)
	return columns, data

def get_columns():
	columns = [
		_("Employee ID") + ":Data:80",_("Employee Name") + ":Data:200",_("Department") + ":Data:200",_("Designation") + ":Data:200",_("Employee Category") + ":Data:150",_("OT") + ":Data:150",_("Comp-off used") + ":Data:100",
		_("OT Balance") + ":Data:120"
	]
	return columns

def get_ot(filters):
    data = []

    if filters.from_date and filters.to_date:
        conditions = {
            'from_date': ('between', (filters.from_date, filters.to_date)),
            'to_date': ('between', (filters.from_date, filters.to_date))
        }

        if filters.employee:
            conditions['employee'] = filters.employee

        otbalance = frappe.get_all(
            "OT Balance",
            filters=conditions,
            fields=[
                "employee",
                "employee_name",
                "department",
				"designation",
				'employee_category',
                "sum(total_ot_hours) as total_ot_hours",
                "sum(comp_off_used) as comp_off_used",
                "sum(ot_balance) as ot_balance",
            ],
            group_by="employee"
        )

        for ot in otbalance:
            row = [ot.employee, ot.employee_name, ot.department,ot.designation,ot.employee_category, ot.total_ot_hours, ot.comp_off_used, ot.ot_balance]
            data.append(row)

    return data

